src/stan/mcmc/chains.hpp

```diff
@@ -292,6 +291,91 @@ namespace stan {
         return ess;
       }
 
+      /**
+       * Returns the effective sample size for the specified parameter
+       * across all kept samples.
+       *
+       * The implementation is close to the effective sample size
+       * description in BDA3 (p. 286-287).  See more details in Stan
+       * reference manual section "Effective Sample Size".
+       *
+       * Current implementation takes the minimum number of samples
+       * across chains as the number of samples per chain.
+       *
+       * @param std::vector<std::vector<double> > samples
+       * @return
+       */
+      double
+      effective_sample_size(const std::vector<std::vector<double> >& samples)
+        const {
+        int chains = samples.size();
+
+        // need to generalize to each jagged samples per chain
+        int n_samples = samples[0].size();
+        for (int chain = 1; chain < chains; chain++) {
+          n_samples = std::min(n_samples,
+                               static_cast<int>(samples[chain].size()));
+        }
+
+        std::vector<std::vector<double> > acov(chains);
+        for (int chain = 0; chain < chains; chain++) {
+          stan::math::autocovariance(samples[chain], acov[chain]);
+        }
+
+        std::vector<double> chain_mean(chains);
+        std::vector<double> chain_var(chains);
+        for (int chain = 0; chain < chains; chain++) {
+          double n_kept_samples = num_kept_samples(chain);
+          chain_mean[chain] = stan::math::mean(samples[chain]);
+          chain_var[chain] = acov[chain][0]*n_kept_samples/(n_kept_samples-1);
+        }
+
+        double mean_var = stan::math::mean(chain_var);
+        double var_plus = mean_var*(n_samples-1)/n_samples;
+        if (chains > 1)
+          var_plus += stan::math::variance(chain_mean);
+        std::vector<double> rho_hat_t(n_samples);
+        std::fill(rho_hat_t.begin(), rho_hat_t.end(), 0);
```

Comment:
You can just use the initializing constructor syntax here, which will also be more efficient.  So that'd just be:

```
std::vector<double> rho_hat_t(n_samples, 0);
```